pkg/jsonexp/expressions_test.go

```diff
@@ -0,0 +1,551 @@
+	assert.NilError(t, err)
+	assert.Check(t, ok)
+
+	// false && nil
+	exp = &And{
+		Left: Pattern{
+			Selector: "str",
+			Operator: EqualOperator,
+			Value:    "wrong-value",
+		},
+	}
+	ok, err = exp.Matches(testJsonData)
+	assert.NilError(t, err)
+	assert.Check(t, !ok)
+
+	// nil && false
+	exp = &And{
+		Right: Pattern{
+			Selector: "str",
+			Operator: EqualOperator,
+			Value:    "wrong-value",
+		},
+	}
+	ok, err = exp.Matches(testJsonData)
+	assert.NilError(t, err)
+	assert.Check(t, !ok)
+}
+
+func TestEmptyAnd(t *testing.T) {
+	// nil && nil
+	exp := &And{}
+	ok, err := exp.Matches(testJsonData)
+	assert.NilError(t, err)
+	assert.Check(t, ok)
+}
+
+func TestOr(t *testing.T) {
+	// true || true
+	exp := &Or{
+		Left: Pattern{
+			Selector: "str",
+			Operator: EqualOperator,
+			Value:    "my-value",
+		},
+		Right: Pattern{
+			Selector: "int",
+			Operator: EqualOperator,
+			Value:    "123",
+		},
+	}
+	ok, err := exp.Matches(testJsonData)
+	assert.NilError(t, err)
+	assert.Check(t, ok)
+
+	// false || true
+	exp = &Or{
+		Left: Pattern{
+			Selector: "str",
+			Operator: EqualOperator,
+			Value:    "wrong-value",
+		},
+		Right: Pattern{
+			Selector: "int",
+			Operator: EqualOperator,
+			Value:    "123",
+		},
+	}
+	ok, err = exp.Matches(testJsonData)
+	assert.NilError(t, err)
+	assert.Check(t, ok)
+
+	// true || false
+	exp = &Or{
+		Left: Pattern{
+			Selector: "str",
+			Operator: EqualOperator,
+			Value:    "my-value",
+		},
+		Right: Pattern{
+			Selector: "int",
+			Operator: EqualOperator,
+			Value:    "wrong-value",
+		},
+	}
+	ok, err = exp.Matches(testJsonData)
+	assert.NilError(t, err)
+	assert.Check(t, ok)
+
+	// false || false
+	exp = &Or{
+		Left: Pattern{
+			Selector: "str",
+			Operator: EqualOperator,
+			Value:    "wrong-value",
+		},
+		Right: Pattern{
+			Selector: "int",
+			Operator: EqualOperator,
+			Value:    "wrong-value",
+		},
+	}
+	ok, err = exp.Matches(testJsonData)
+	assert.NilError(t, err)
+	assert.Check(t, !ok)
+}
+func TestOneBranchOr(t *testing.T) {
+	// true || nil
+	exp := &Or{
+		Left: Pattern{
+			Selector: "str",
+			Operator: EqualOperator,
+			Value:    "my-value",
+		},
+	}
+	ok, err := exp.Matches(testJsonData)
+	assert.NilError(t, err)
+	assert.Check(t, ok)
+
+	// nil || true
+	exp = &Or{
+		Right: Pattern{
+			Selector: "str",
+			Operator: EqualOperator,
+			Value:    "my-value",
+		},
+	}
+	ok, err = exp.Matches(testJsonData)
+	assert.NilError(t, err)
+	assert.Check(t, ok)
+
+	// false || nil
+	exp = &Or{
+		Left: Pattern{
+			Selector: "str",
+			Operator: EqualOperator,
+			Value:    "wrong-value",
+		},
+	}
+	ok, err = exp.Matches(testJsonData)
+	assert.NilError(t, err)
+	assert.Check(t, !ok)
+
+	// nil || false
+	exp = &Or{
+		Right: Pattern{
+			Selector: "str",
+			Operator: EqualOperator,
+			Value:    "wrong-value",
+		},
+	}
+	ok, err = exp.Matches(testJsonData)
+	assert.NilError(t, err)
+	assert.Check(t, !ok)
+}
+
+func TestEmptyOr(t *testing.T) {
```

Comment:
perhaps being stupid, but not getting why an empty `or` returns false but an empty `and` returns true?